### PR TITLE
Fix & improve perf for ROCm

### PR DESCRIPTION
**Pull request by chhwang**

### CODE REVIEW

File: include/mscclpp/fifo_device.hpp [hunk 1]
```diff
@@ -70,9 +70,9 @@ struct FifoDeviceHandle {
 #if defined(MSCCLPP_DEVICE_CUDA)
     asm volatile("st.global.relaxed.sys.v2.u64 [%0], {%1,%2};" ::"l"(triggerPtr), "l"(trigger.fst), "l"(trigger.snd));
 #else   // !defined(MSCCLPP_DEVICE_CUDA)
-    // TODO: both atomic and clang built-ins are buggy here
-    triggerPtr->fst = trigger.fst;
-    triggerPtr->snd = trigger.snd;
+    // store snd no later than fst.
+    atomicStore(&(triggerPtr->snd), trigger.snd, memoryOrderRelaxed);
```
Comment (saeedmaleki) [inline include/mscclpp/fifo_device.hpp:74]: why atomic?

File: test/mscclpp-test/allreduce_test.cu [hunk 2]
```diff
@@ -861,9 +956,15 @@ __global__ void allreduce4(int* buff, int* scratch, int rank, int nRanksPerNode,
 }
 
 __global__ void allreduce5(int* buff, int rank, int nRanksPerNode, int worldSize, size_t nelems) {
+#if defined(__HIP_PLATFORM_AMD__) && (__HIP_PLATFORM_AMD__ == 1)
+  localReduceScatterSm3(buff, rank, nRanksPerNode, nelems / worldSize, nelems / worldSize, gridDim.x);
```
Comment (saeedmaleki) [inline test/mscclpp-test/allreduce_test.cu:960]: do you have perf numbers for these?

**Review by saeedmaleki — APPROVED** [review]

**Review by Binyang2014 — APPROVED** [review]

File: include/mscclpp/fifo_device.hpp [hunk 3]
```diff
@@ -70,9 +70,9 @@ struct FifoDeviceHandle {
 #if defined(MSCCLPP_DEVICE_CUDA)
     asm volatile("st.global.relaxed.sys.v2.u64 [%0], {%1,%2};" ::"l"(triggerPtr), "l"(trigger.fst), "l"(trigger.snd));
 #else   // !defined(MSCCLPP_DEVICE_CUDA)
-    // TODO: both atomic and clang built-ins are buggy here
-    triggerPtr->fst = trigger.fst;
-    triggerPtr->snd = trigger.snd;
+    // store snd no later than fst.
```
[the remark this answers is not in the document]
Comment (saeedmaleki) [reply include/mscclpp/fifo_device.hpp:73]: why does it matter? for the proxy we check whether both of them are non-zero.

File: include/mscclpp/fifo_device.hpp (hunk 1 above)
Comment (chhwang) [reply include/mscclpp/fifo_device.hpp:74]: I was worried of optimize-out by the compiler. It doesn't affect the latency as well.

File: test/mscclpp-test/allreduce_test.cu (hunk 2 above)
Comment (chhwang) [reply test/mscclpp-test/allreduce_test.cu:960]: yes, numbers are reported last week.

File: include/mscclpp/fifo_device.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (chhwang) [reply include/mscclpp/fifo_device.hpp:73]: Currently it does matter but not sure of the reason yet. I'm fixing the issue first with minimal changes. Will create an issue for this and revisit

**Pull request merged**